File: src/update/update-cli.ts
/**
 * ⚠️ LEGACY TypeScript client — frozen. Fix the Rust client instead.
 *
 * The shipping `climon` *client* is the Rust workspace under `rust/` (crates
 * `climon-cli`, `climon-session`, `climon-pty`, `climon-store`, `climon-config`,
 * `climon-remote`, `climon-install`, `climon-update`, …). This module belongs to
 * the legacy Bun/TypeScript client, kept only for local development and the Bun
 * test suite. Do NOT add features or fix client bugs here — make all client
 * changes in the Rust crates. (The Bun dashboard *server* under `src/server*`
 * and `src/web/` is NOT legacy and is still maintained.)
 */
import { dirname } from "node:path";
import { readGlobalConfigSetting } from "../config.js";
import { t } from "../i18n/t.js";
import { VERSION } from "../version.js";
import { DEFAULT_MANIFEST_URL } from "./check.js";
import { fetchManifest } from "./manifest.js";
import { UPDATE_PUBLIC_KEY_B64 } from "./pubkey.js";
import { clearAvailableVersion } from "./state.js";
import { runUpdateCommand } from "./update-cmd.js";

/**
 * Reads the shared decryption password from the global config (per-machine;
 * never shadowed by a project-local config). Returns undefined when unset.
 */
export function getConfiguredUpdatePassword(
  env: NodeJS.ProcessEnv = process.env
): string | undefined {
  const value = readGlobalConfigSetting("update.password", env);
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

/** `climon update` entrypoint: resolves install dir and applies an update. */
export async function runUpdateCli(_argv: string[]): Promise<number> {
  const installDir = dirname(process.execPath);
  try {
    const manifest = await fetchManifest(DEFAULT_MANIFEST_URL);
    const result = await runUpdateCommand({
      installDir,
      currentVersion: VERSION,
      manifest,
      publicKeyB64: UPDATE_PUBLIC_KEY_B64,
      decryptPassword: getConfiguredUpdatePassword(process.env),
    });
    if (result.status === "updated") clearAvailableVersion(process.env);
    return result.status === "verify-failed" ||
      result.status === "decrypt-failed" ||
      result.status === "no-artifact"
      ? 1
      : 0;
  } catch (err) {
    process.stderr.write(t("update.failed", { detail: (err as Error).message }) + "\n");
    return 1;
  }
}

File: rust/climon-update/src/update_cli.rs
//! `climon update` entrypoint. Port of `src/update/update-cli.ts`.

use std::path::Path;

use climon_config::config::Env;

use crate::check::DEFAULT_MANIFEST_URL;
use crate::manifest::fetch_manifest;
use crate::pubkey::UPDATE_PUBLIC_KEY_B64;
use crate::state::clear_available_version;
use crate::update_cmd::{run_update_command, UpdateCommandOptions, UpdateStatus};
use crate::version::VERSION;

/// `climon update` entrypoint: resolves the install dir and applies an update.
/// Returns the process exit code (0 success/up-to-date/deferred; 1 on failure).
pub fn run_update_cli(_argv: &[String], env: &Env) -> i32 {
    let exe = match std::env::current_exe() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("climon update failed: {e}");
            return 1;
        }
    };
    let install_dir = exe.parent().unwrap_or(Path::new(".")).to_path_buf();

    let manifest = match fetch_manifest(DEFAULT_MANIFEST_URL) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("climon update failed: {e}");
            return 1;
        }
    };

    let opts = UpdateCommandOptions {
        install_dir: &install_dir,
        current_version: VERSION,
        manifest: &manifest,
        public_key_b64: UPDATE_PUBLIC_KEY_B64,
        platform: crate::manifest::current_node_platform(),
        arch: crate::manifest::current_node_arch(),
    };

    let result = match run_update_command(&opts, &mut |s| print!("{s}")) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("climon update failed: {e}");
            return 1;
        }
    };

    if result.status == UpdateStatus::Updated {
        clear_available_version(env);
    }

    match result.status {
        UpdateStatus::VerifyFailed | UpdateStatus::NoArtifact => 1,
        _ => 0,
    }
}
